modes.go

```diff
@@ -287,23 +290,123 @@ func RunTest(resultChannel chan Result, workload WorkloadGenerator, rateLimiter
 	resultChannel <- *rb.FullResult
 }
 
-func generateData(pk int64, ck int64, size int64) []byte {
+const (
+	generatedDataHeaderSize int64 = 24
+	generatedDataMinSize int64 = generatedDataHeaderSize + 33
+)
+
+func GenerateData(pk int64, ck int64, size int64) []byte {
+	if !validateData {
+		return make([]byte, size);
+	}
+
+	buf := new(bytes.Buffer)
+
+	if size < generatedDataHeaderSize {
+		binary.Write(buf, binary.LittleEndian, int8(size))
+		binary.Write(buf, binary.LittleEndian, pk ^ ck)
+	} else {
+		binary.Write(buf, binary.LittleEndian, size)
+		binary.Write(buf, binary.LittleEndian, pk)
+		binary.Write(buf, binary.LittleEndian, ck)
+		if size < generatedDataMinSize {
+			for i := generatedDataHeaderSize; i < size; i++ {
+				binary.Write(buf, binary.LittleEndian, int8(0))
+			}
+		} else {
+			payload := make([]byte, size - generatedDataHeaderSize - sha256.Size)
+			rand.Read(payload)
+			csum := sha256.Sum256(payload)
+			binary.Write(buf, binary.LittleEndian, payload)
+			binary.Write(buf, binary.LittleEndian, csum)
+		}
+	}
+
 	value := make([]byte, size)
-	if validateData {
-		dataPattern := strconv.FormatInt(pk*ck*(pk+ck), 10)
-		dataLen := len(dataPattern)
-		cnt := int(size) / dataLen
-		tail := int(size) % dataLen
-		var data string
-		if cnt > 0 {
-			data = strings.Repeat(dataPattern, cnt)
+	copy(value, buf.Bytes())
+	return value
+}
+
+func ValidateData(pk int64, ck int64, data []byte) error {
+	if !validateData {
+		return nil
+	}
+
+	buf := bytes.NewBuffer(data)
+	size := int64(buf.Len())
+
+	var storedSize int64
+	if size < generatedDataHeaderSize {
+		var storedSizeCompact int8
+		err := binary.Read(buf, binary.LittleEndian, &storedSizeCompact)
+		if err != nil {
+			return errors.Wrap(err, "failed to validate data, cannot read size from value")
 		}
-		if tail > 0 {
-			data += dataPattern[:tail]
+		storedSize = int64(storedSizeCompact)
+	} else {
+		err := binary.Read(buf, binary.LittleEndian, &storedSize)
+		if err != nil {
+			return errors.Wrap(err, "failed to validate data, cannot read size from value")
 		}
-		copy(value, []byte(data))
 	}
-	return value
+
+	if size != storedSize {
+		return errors.Errorf("actual size of value (%d) doesn't match size stored in value (%d)", size, storedSize)
+	}
+
+	// There is no random payload for sizes < minFullSize
+	if size < generatedDataMinSize {
+		expectedBuf := GenerateData(pk, ck, size)
+		if !bytes.Equal(buf.Bytes(), expectedBuf) {
+			return errors.Errorf("actual value doesn't match expected value:\nexpected: %x\nactual: %x", expectedBuf, buf.Bytes())
+		}
+	} else {
```

Comment:
nit: could we return here and avoid else?

Reply:
Good idea.